client/client.go

```diff
@@ -1247,6 +1289,207 @@ func (c *Client) runAllocs(update *allocUpdates) {
 	}
 }
 
+// blockForRemoteAlloc blocks until the previous allocation of an allocation has
+// been terminated and migrates the snapshot data
+func (c *Client) blockForRemoteAlloc(alloc *structs.Allocation) {
+	c.logger.Printf("[DEBUG] client: blocking alloc %q for previous allocation %q", alloc.ID, alloc.PreviousAllocation)
+
+	// Removing the allocation from the set of allocs which are currently
+	// undergoing migration
+	defer func() {
+		c.migratingAllocsLock.Lock()
+		delete(c.migratingAllocs, alloc.ID)
+		c.migratingAllocsLock.Unlock()
+	}()
+
+	var err error
+	var prevAlloc *structs.Allocation
+	var index uint64
+	for {
+		// Stopping if the client is shutting down.
+		if c.shutdown {
+			return
+		}
+
+		// get the remote allocation.
+		prevAlloc, index, err = c.getAllocation(alloc.PreviousAllocation, index)
+		if err != nil {
+			c.logger.Printf("[ERROR] client: error getting previous allocation %q for alloc %q: %v",
+				alloc.PreviousAllocation, alloc.ID, err)
+			return
+		}
+
+		// If the previous allocation has been terminated then we stop.
+		if prevAlloc == nil || prevAlloc.Terminated() {
+			break
+		}
+		continue
+	}
+
+	var prevAllocDir *allocdir.AllocDir
+	if prevAlloc != nil {
+		tg := alloc.Job.LookupTaskGroup(alloc.TaskGroup)
+
+		// Migrate the data from remote node if the disk is sticky and migrate
+		// is turned on
+		if tg != nil && tg.EphemeralDisk.Sticky && tg.EphemeralDisk.Migrate {
+			// Create new alloc dir
+			if node, err := c.getNode(prevAlloc.NodeID); err == nil && node.Status != structs.NodeStatusDown {
+				// Create the previous alloc dir
+				pathToAllocDir := filepath.Join(c.config.AllocDir, prevAlloc.ID)
+				if err := os.MkdirAll(pathToAllocDir, 0777); err != nil {
+					c.logger.Printf("[ERR] client: error creating previous allocation dir: %v", err)
+				}
+
+				// Get the snapshot
+				url := fmt.Sprintf("http://%v/v1/client/allocation/%v/snapshot", node.HTTPAddr, prevAlloc.ID)
+				resp, err := http.Get(url)
+				if err != nil {
+					c.logger.Printf("[ERR] client: error getting snapshot: %v", err)
+				}
+				tr := tar.NewReader(resp.Body)
+				defer resp.Body.Close()
+
+				buf := make([]byte, 32)
+				for {
+					// If the client is shutting down then we return
```

Comment:
```
// Get once
stopMigrating, ok := c.migratingAllocs[alloc.ID]
if !ok {
   return err since this is a pre-condition of this function
}

select {
case <-c.shutdownCh:
   // log
   return nil
case <- stopMigrating
   log
   return nil
}
```